### PR TITLE
Support reloading games while preserving windows and other resources

### PR DESCRIPTION
**Pull request by bjorn3**
<details><summary>gif</summary>

</details>

### CODE REVIEW

File: crates/bevy_ecs/src/resource/resources.rs [hunk 1]
```diff
@@ -320,6 +320,29 @@ impl Resources {
         })
     }
 
+    pub fn take_global_only_resource<T: Resource>(&mut self) -> Option<T> {
```
Comment (cart) [inline crates/bevy_ecs/src/resource/resources.rs:323]: This name/impl definitely makes sense in the context of the current implementation, but it makes less sense from a user facing perspective. Ideally we could have `resources.take::<T>()`, which takes a global, and `resources.take_local::<T>(index)` which takes a local.

**Review by cart — COMMENTED** [review]
This is so cool :smile: 

I think I'm going to hold off on merging as I think there might be synergies between `bevy_persist` and the new `bevy_reflect` crate im wrapping up. I'm hoping it will cut down on some of the boilerplate / special casing required for "persisted resources".

**Comment by cart** [comment]
It would also be cool if we could unify `bevy_dynamic_plugin` and `bevy_persist` conceptually, but I totally understand that "reloading individual plugins" is very different from reloading the whole app (and as you pointed out on the discord, hot-reloading plugins has lots of unsafety potential).

**Comment by steelx** [comment]
wow this looks like Flutter hot-reload. hope it does comes true 💯

**Comment by alice-i-cecile** [comment]
@bjorn3 is this ready for final review before merging?

**Comment by bjorn3** [comment]
No, it is not. The more I think about it, the more I think that at least window preservation should happen through an external process (in the future the editor, for now probably more something like a stub), rather than what this uses with in-process hot reloading. This external process could also save all state. This is also a lot more safe and much less likely to break due to a dylib not unloading for whatever reason.

**Comment by bjorn3** [comment]
I don't think I will be working on this any time soon. As such I have marked it as S-Adopt-Me.

**Comment by NthTensor** [comment]
Heads up, we are now trying to close PRs that arn't being actively worked on.

**Pull request closed without merging**